llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -7436,6 +7437,110 @@ static bool simplifySwitchOfCmpIntrinsic(SwitchInst *SI, IRBuilderBase &Builder,
+  // same.
+
+  // Find BBs that are candidates for simplification.
+  SmallPtrSet<BasicBlock *, 8> BBs;
+  for (auto &Case : SI->cases()) {
+    BasicBlock *BB = Case.getCaseSuccessor();
+
+    // FIXME: This case needs some extra care because the terminators other than
+    // SI need to be updated.
+    if (!BB->hasNPredecessors(1))
+      continue;
+
+    // FIXME: Relax that the terminator is a BranchInst by checking for equality
+    // on other kinds of terminators.
+    Instruction *T = BB->getTerminator();
+    if (T && isa<BranchInst>(T))
+      BBs.insert(BB);
+  }
+
+  auto IsBranchEq = [](BranchInst *A, BranchInst *B) {
+    if (A->isConditional() != B->isConditional())
+      return false;
+
+    if (A->isConditional()) {
+      // If the conditions are instructions, check equality up to commutativity.
+      // Otherwise, check that the two Values are the same.
+      Value *AC = A->getCondition();
+      Value *BC = B->getCondition();
+      auto *ACI = dyn_cast<Instruction>(AC);
+      auto *BCI = dyn_cast<Instruction>(BC);
+      if ((ACI && BCI && !areIdenticalUpToCommutativity(ACI, BCI)) && AC != BC)
+        return false;
+    }
+
+    if (A->getNumSuccessors() != B->getNumSuccessors())
+      return false;
+
+    for (unsigned I = 0; I < A->getNumSuccessors(); ++I) {
+      BasicBlock *ASucc = A->getSuccessor(I);
+      if (ASucc != B->getSuccessor(I))
+        return false;
+      // Need to check that PHIs in sucessors have matching values
+      for (PHINode &Phi : ASucc->phis())
+        if (Phi.getIncomingValueForBlock(A->getParent()) !=
+            Phi.getIncomingValueForBlock(B->getParent()))
+          return false;
+    }
+
+    return true;
+  };
+
+  auto IsBBEqualTo = [&IsBranchEq](BasicBlock *A, BasicBlock *B) {
+    // FIXME: Support more than just a single BranchInst. One way we could do
+    // this is by taking a hashing approach.
+    if (A->size() != 1 || B->size() != 1)
+      return false;
+
+    return IsBranchEq(cast<BranchInst>(A->getTerminator()),
+                      cast<BranchInst>(B->getTerminator()));
+  };
+
+  // Construct a map from candidate basic block to an equivalent basic block
+  // to replace it with. All equivalent basic blocks should be replaced with
+  // the same basic block. To do this, if there is no equivalent BB in the map,
+  // then insert into the map BB -> BB. Otherwise, we should check only elements
+  // in the map for equivalence to ensure that all equivalent BB get replaced
+  // by the BB in the map. Replacing BB with BB has no impact, so we skip
+  // a call to setSuccessor when we do the actual replacement.
+  DenseMap<BasicBlock *, BasicBlock *> ReplaceWith;
+  for (BasicBlock *BB : BBs) {
+    bool Inserted = false;
+    for (auto KV : ReplaceWith) {
+      if (IsBBEqualTo(BB, KV.first)) {
+        ReplaceWith[BB] = KV.first;
+        Inserted = true;
+        break;
+      }
+    }
+    if (!Inserted)
+      ReplaceWith[BB] = BB;
+  }
```

Comment:
I think the way to implement this is with `DenseMap<BasicBlock *, SmallVector<BasicBlock *>>`. If a given BB is not in `ReplaceWith`, you add it w/ empty list. If it is, add to list of existing entry. You then iterate over all key BBs in `ReplaceWith` and replace all items in the `SmallVector` with the key BB.

In this case your hash function can just be zero (or maybe `BB->size()`) and `isBBEqualTo` is your equals function.

Reply:
Could you please help me understand why you propose this as a suggestion? Do you think the current approach is incorrect, slow, uses too much memory, or is overcomplicated?

I tried to implement your suggestion, but I'm definitely misunderstanding what you mean.

This is what I *think* you are saying:

```
DenseMap<BasicBlock *, SmallVector<BasicBlock *>> ReplaceWith;
// Snip: create a map where values are the basic blocks that should 
// be replaced with the key basic block. Must implement
// DenseMapInfo to make it work.

// Iterate over all key BBs in ReplaceWith and replace all items in the
// SmallVector with the key BB.
for (auto KV : ReplaceWith) 
  for (BasicBlock *BB : KV.second)
    for (&Case : SI->cases())
      if (Case.getCaseSuccessor() = BB)
        Case.setSuccessor(KV.first);
```

Again, there is no way to get BasicBlock -> Cases mapping without looping over all the cases and checking that the CaseSuccessor is equal to BB. This approach is less performant because we have to iterate over cases for every (KV, BB) pair. In the current approach, we have O(1) lookup  -> ReplaceWith[Case.getCaseSuccessor].

We could change the loop to:

```
for (&Case : SI->cases()) {
  for (auto KV : ReplaceWith)
    for (BasicBlock *BB : KV.second)
      if (Case.getCaseSuccessor() == BB)
        Case.setSuccessor(KV.first);
```

In this case, we still have worse performance since we don't have O(1) lookup.

Reply:
I am worried about the O(n^2).

I think you can do it by making the map use `CaseHandle` directly i.e:
```
DenseMap<CaseHandle *, SmallVector<CaseHandle *>, CustomKeyInfo> ReplaceWith;
for (auto Case : SI->cases()) {
  auto Res = ReplaceWith.emplace(&Case, {});
  if (Res.second)
    continue;
  Res.first->second.push_back(&Case);
}
...
for (auto KV : ReplaceWith) {
  for (auto ToReplace : KV.second) {
    ToReplace->setSuccessor(KV.first);
  }
}

```

Where `CustomKeyInfo` is something along the lines of :
```
struct CustomKeyInfo {
  static inline CaseHandle *getEmptyKey() { return nullptr; }
  static inline CaseHandle *getTombstoneKey() {
    return reinterpret_cast<CaseHandle *>(-1ULL);
  }
	// I would lookup what we do in EarlyCSE
  static unsigned getHashValue(const CaseHandle* &Val) {
	  // TODO
  }
	static bool isEqual(const CaseHandle* &LHS, const CaseHandle* &RHS) {
		return isBBEqualTo(LHS->getCaseSuccessor(), RHS->getCaseSuccessor());
	}

}
```

Reply:
I agree we need a hash-based O(n) method to merge these blocks.

Compile-time impact:
```
Top 5 regressions:
  llvm/TokenKinds.cpp.ll 151977954 -> 792848805 +421.69%
  cvc5/kind.cpp.ll 230032379 -> 1046824473 +355.08%
  llvm/ASTCommon.cpp.ll 519331147 -> 2229199828 +329.24%
  llvm/X86EncodingOptimization.cpp.ll 424207707 -> 1707051386 +302.41%
  rust-analyzer-rs/4ifo5x52byu175vr.ll 177537147 -> 557775939 +214.17%
```

Reply:
Thank you for your suggestions! It is a great idea to use CaseHandle directly. I have updated the patch with this improved implementation. It is way better!

@dtcxzyw just curious, what tool did you measure the compile time impact? Seems like it would be useful to use in my own workflow.

Reply:
He used: https://github.com/dtcxzyw/llvm-opt-benchmark/
nikic also supports http://llvm-compile-time-tracker.com/.

At the moment there are some issues with the compile-times from https://github.com/dtcxzyw/llvm-opt-benchmark/ but a 400% regression is clearly an issue.

Reply:
I agree that a 400% regression is clearly an issue. I'm hoping that my commits from today resolve those problems. Do I have the power to trigger a run of llvm-opt-benchmark or llvm-compile-time-tracker? Gave a look around the links you shared with me, but I didn't see any instructions, and I couldn't find my commits on http://llvm-compile-time-tracker.com/.

Reply:
See: https://github.com/dtcxzyw/llvm-opt-benchmark/issues/1312#issuecomment-2450627194

Reply:
http://llvm-compile-time-tracker.com/compare.php?from=6f973fd4ab18ff58689e83383190ed4767c2a7dd&to=9e9afc2783de2f5d323c84e1bd8c2284322bce91&stat=instructions%3Au

Looks like still notable regressions.

Reply:
I am seeing regressions <= 1%. What is the threshold for notable?

Reply:
Are we able to give this PR another run? I made some changes that I hope can sway the performance in our favor.

Reply:
See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1587.
The result looks worse :(
```
Top 5 regressions:
  llvm/TokenKinds.cpp.ll 151977954 -> 1946532443 +1180.80%
  llvm/X86EncodingOptimization.cpp.ll 424207707 -> 4860819739 +1045.86%
  llvm/ASTCommon.cpp.ll 518542127 -> 5935397105 +1044.63%
  cvc5/kind.cpp.ll 229757441 -> 2437197537 +960.77%
  libquic/net_errors.cc.ll 121648675 -> 1077626563 +785.85%
```